win32/src/threadpool_win32.c

```diff
@@ -346,75 +350,96 @@ static void threadpool_internal_cancel_timer_and_wait(PTP_TIMER tp_timer)
+            LogError("failure in THANDLE_MALLOC(THREADPOOL_TIMER)(threadpool_timer_dispose=%p)",
+                    threadpool_timer_dispose);
         }
         else
         {
+            result->timer = NULL;
+
             /* Codes_SRS_THREADPOOL_WIN32_42_006: [ threadpool_timer_start shall call CreateThreadpoolTimer to schedule execution the callback while passing to it the on_timer_callback function and the newly created context. ]*/
-            PTP_TIMER tp_timer = CreateThreadpoolTimer(on_timer_callback, timer_temp, &threadpool_ptr->tp_environment);
+            PTP_TIMER tp_timer = CreateThreadpoolTimer(on_timer_callback, result, &threadpool_ptr->tp_environment);
             if (tp_timer == NULL)
             {
                 /* Codes_SRS_THREADPOOL_WIN32_42_008: [ If any error occurs, threadpool_timer_start shall fail and return a non-zero value. ]*/
                 LogError("CreateThreadpoolTimer failed");
-                result = MU_FAILURE;
             }
             else
             {
-                timer_temp->timer = tp_timer;
-                timer_temp->work_function = work_function;
-                timer_temp->work_function_context = work_function_context;
+                result->timer = tp_timer;
+
+                /* Codes_SRS_THREADPOOL_WIN32_07_002: [ threadpool_timer_start shall initialize the lock guarding the timer state by calling srw_lock_ll_init. ]*/
+                if (srw_lock_ll_init(&result->timer_lock) != 0)
+                {
+                    /* Codes_SRS_THREADPOOL_WIN32_42_008: [ If any error occurs, threadpool_timer_start shall fail and return NULL. ]*/
+                    LogError("srw_lock_ll_init failed");
+                }
+                else
+                {
+                    result->work_function = work_function;
+                    result->work_function_context = work_function_context;
 
-                /* Codes_SRS_THREADPOOL_WIN32_42_007: [ threadpool_timer_start shall call SetThreadpoolTimer, passing negative start_delay_ms as pftDueTime, timer_period_ms as msPeriod, and 0 as msWindowLength. ]*/
-                threadpool_internal_set_timer(tp_timer, start_delay_ms, timer_period_ms);
+                    /* Codes_SRS_THREADPOOL_WIN32_42_007: [ threadpool_timer_start shall call SetThreadpoolTimer, passing negative start_delay_ms as pftDueTime, timer_period_ms as msPeriod, and 0 as msWindowLength. ]*/
+                    threadpool_internal_set_timer(tp_timer, start_delay_ms, timer_period_ms);
 
-                /* Codes_SRS_THREADPOOL_WIN32_42_009: [ threadpool_timer_start shall return the allocated handle in timer_handle. ]*/
-                *timer_handle = timer_temp;
-                timer_temp = NULL;
+                    /* Codes_SRS_THREADPOOL_WIN32_42_009: [ threadpool_timer_start shall return the allocated handle in timer_handle. ]*/
+                    //THANDLE_INITIALIZE_MOVE(THREADPOOL_TIMER)(timer_handle, &timer_temp);

```

Comment:
>//THANDLE_INITIALIZE_MOVE(THREADPOOL_TIMER)(timer_handle, &timer_temp); [](http://example.com/codeflow?start=20&length=71)

Commented out code detected, not sure what is happening with this one.

Reply:
forgot to remove, the timer need to be returned already set to result. deleted this line